Docs: add apache amoro(incubating) with iceberg (#11965) (#11966)
apache/iceberg: docs/docs/amoro.md

---
title: "Apache Amoro"
---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Apache Amoro With Iceberg

> **[Apache Amoro(incubating)](https://amoro.apache.org)** is a Lakehouse management system built on open data lake formats. Working with compute engines including Flink, Spark, and Trino, Amoro brings pluggable and

Reviewer: Does "pluggable and" means pluggable and Table Maintenance/AMS?

Author: Includes pluggable monitoring plugins and table maintenance merge scheduling strategy.Cann‘t list them all in doc.

Reviewer 2: It reads strange having "pluggable" side-by-side with "Table Maintenance". Why not using the words from [Amoro README](https://github.com/apache/amoro/blob/master/README.md)

> Amoro brings pluggable and self-managed features for Lakehouse to provide out-of-the-box data warehouse experience ...

Reviewer 2: @czy006 please check this comment.

**[Table Maintenance](https://amoro.apache.org/docs/latest/self-optimizing/)** features for a Lakehouse to provide out-of-the-box data warehouse experience, and helps data platforms or products easily build infra-decoupled, stream-and-batch-fused and lake-native architecture.

> **AMS(Amoro MetaStore)** is Amoro Management Service provides Lakehouse management features, like self-optimizing, data expiration, etc. It also provides a unified catalog service for all compute engines, which can also be combined with existing metadata services.

Reviewer 3: Should this rather say `AMS (Amore MetaStore) is a management service that provides Lakehouse management features...`? Otherwise the first part of the sentence is confusing

Author: Thank you for correcting this incorrect description. AMS is a management service (sometime confuse by myself...) fix it

> **AMS(Amoro Management Service)**  provides Lakehouse management features, like self-optimizing, data expiration, etc. It also provides a unified catalog service for all compute engines, which can also be combined with existing metadata services like HMS(Hive Metastore).

Reviewer 2: nit: extra space before "provides"

> # Auto Self-optimizing

Reviewer 2: These headers need to be at second level.

> **AMS(Amoro Management Service)**  provides Lakehouse management features, like self-optimizing, data expiration, etc. It also provides [REST catalog](https://iceberg.apache.org/concepts/catalog/#decoupling-using-the-rest-catalog) service for all compute engines, which can also be combined with existing metadata services like HMS(Hive Metastore).

Reviewer 3: ```suggestion
**AMS(Amoro Management Service)**  provides Lakehouse management features, like self-optimizing, data expiration, etc. It also provides a unified catalog service for all compute engines, which can also be combined with existing metadata services like HMS(Hive Metastore).
```

I feel like mentioning REST and linking to Iceberg here is rather confusing and probably not what's being intended here?

Lakehouse is characterized by its openness and loose coupling, with data and files maintained by users through various engines. While this
architecture appears to be well-suited for T+1 scenarios, as more attention is paid to applying Lakehouse to streaming data warehouses and real-time
analysis scenarios, challenges arise. For example:

- Streaming writes bring a massive amount of fragment files
- CDC ingestion and streaming updates generate excessive redundant data
- Using the new data lake format leads to orphan files and expired snapshots.

These issues can significantly affect the performance and cost of data analysis. Therefore, Amoro has introduced a Self-optimizing mechanism to
create an out-of-the-box Streaming Lakehouse management service that is as user-friendly as a traditional database or data warehouse. Self-optimizing involves various procedures such as file compaction, deduplication, and sorting.

The architecture and working mechanism of Self-optimizing are shown in the figure below:

![Self-optimizing architecture](https://github.com/apache/amoro/blob/master/docs/images/concepts/self-optimizing_arch.png)

> The Optimizer is a component responsible for executing Self-optimizing tasks. It is a resident process managed by [AMS](https://amoro.apache.org/docs/latest/#architecture). AMS is responsible for

Reviewer 2: Why is AMS not linked when it's firstly introduced above?

> The Optimizer is a component responsible for executing Self-optimizing tasks. It is a resident process managed by AMS(Amoro MetaService). AMS is responsible for

Reviewer 3: this says AMS stands for Amore MetaService but a few lines above AMS stands for Amoro Management Service. Which is correct now?

Ideally I think this readme should really only have a few pointers about what Amoro is and point to the official docs in order to avoid frequent updates to this site whenever some functionality/features change btw

detecting and planning Self-optimizing tasks for tables, and then scheduling them to Optimizers for distributed execution in real-time. Finally, AMS
is responsible for submitting the optimizing results. Amoro achieves physical isolation of Optimizers through the Optimizer Group.

The core features of [Amoro's Self-optimizing](https://amoro.apache.org/docs/latest/self-optimizing/) are:

- Automated, Asynchronous and Transparent — Continuous background detecting of file changes, asynchronous distributed execution of optimizing tasks,
  transparent and imperceptible to users
- Resource Isolation and Sharing — Allow resources to be isolated and shared at the table level, as well as setting resource quotas
- Flexible and Scalable Deployment — Optimizers support various deployment methods and convenient scaling

> # Iceberg Format

Reviewer 2: * I'd suggest adding a second-level header `Table Formats` with a third-level header for each format
* The first two paragraphs below are not related to "Format"

> Apache Amoro supports all catalog types supported by Iceberg, including but not limited to: Hadoop, Hive, Glue, JDBC, Nessie, Snowflake, and so on.

Reviewer 3: since this mentions **all**, does that also include the REST catalog?

Author: sure,rest catalog include it.

> Apache Amoro supports all catalog types supported by Iceberg, including common catalog: Hadoop, Hive, Glue, JDBC, Nessie, Snowflake, [Rest Catalog](https://iceberg.apache.org/concepts/catalog/#decoupling-using-the-rest-catalog) and other third-party catalog.

Reviewer 4: ```suggestion
Apache Amoro supports all catalog types supported by Iceberg, including common catalog: [REST](https://iceberg.apache.org/concepts/catalog/#decoupling-using-the-rest-catalog), Hadoop, Hive, Glue, JDBC, Nessie and other third-party catalog.
```

> Amoro supports all storage types supported by Iceberg, including common object store: Hadoop, S3, GCS, ECS, OSS, and so on.

Reviewer 4: I believe Hadoop is not an object store.

At the same time, we also provide a unique form based on Apache Iceberg, including mixed-Iceberg Format and mixed-Hive Format, so that you can quickly upgrade to the iceberg+hive Mixed table while compatible with the original Hive data

## Mixed-Iceberg Format

> [Mixed-Iceberg Format](https://amoro.apache.org/docs/latest/self-optimizing/) Compared with Iceberg format, Mixed-Iceberg format provides more features:

Reviewer 5: The link seems wrong. I guess https://amoro.apache.org/docs/latest/mixed-iceberg-format/ is the correct link.

---
title: "Apache Amoro"
---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Apache Amoro With Iceberg

> **[Apache Amoro(incubating)](https://amoro.apache.org/docs/latest/)** is a Lakehouse management system built on open data lake formats. Working with compute engines including Flink, Spark, and Trino, Amoro brings pluggable and

Reviewer 4: ```suggestion
**[Apache Amoro(incubating)](https://amoro.apache.org/)** is a Lakehouse management system built on open data lake formats. Working with compute engines including Flink, Spark, and Trino, Amoro brings pluggable and
```

> **[Table Maintenance](https://amoro.apache.org/docs/latest/self-optimizing/)** features for Lakehouse to provide out-of-the-box data warehouse experience, and helps data platforms or products easily build infra-decoupled, stream-and-batch-fused and lake-native architecture.

Reviewer 3: ```suggestion
**[Table Maintenance](https://amoro.apache.org/docs/latest/self-optimizing/)** features for a Lakehouse to provide out-of-the-box data warehouse experience, and helps data platforms or products easily build infra-decoupled, stream-and-batch-fused and lake-native architecture.
```

# Auto Self-optimizing

Lakehouse is characterized by its openness and loose coupling, with data and files maintained by users through various engines. While this
architecture appears to be well-suited for T+1 scenarios, as more attention is paid to applying Lakehouse to streaming data warehouses and real-time
analysis scenarios, challenges arise. For example:

- Streaming writes bring a massive amount of fragment files
- CDC ingestion and streaming updates generate excessive redundant data
- Using the new data lake format leads to orphan files and expired snapshots.

These issues can significantly affect the performance and cost of data analysis. Therefore, Amoro has introduced a Self-optimizing mechanism to
create an out-of-the-box Streaming Lakehouse management service that is as user-friendly as a traditional database or data warehouse. The new table

> format is used for this purpose. Self-optimizing involves various procedures such as file compaction, deduplication, and sorting.

Reviewer 3: what does `The new table format is used for this purpose` mean exactly? It suggests to the reader that Amoro introduces a separate table format?

Reviewer 4: I believe it is a streaming-enhanced table format based on Iceberg. The table format essentially extends Iceberg by combining two Iceberg tables.

We can remove this statement as it lacks clarity in the current context.

Author: I want to express that not only Apache Iceberg, but in the future Apache Amoro, any lake format can be applied to this management. I will remove this expression here to avoid misunderstanding as it is a **new format**

The architecture and working mechanism of Self-optimizing are shown in the figure below:

![Self-optimizing architecture](https://github.com/apache/amoro/blob/master/docs/images/concepts/self-optimizing_arch.png)

> The Optimizer is a component responsible for executing Self-optimizing tasks. It is a resident process managed by AMS. AMS is responsible for

Reviewer 5: Is AMS an abbreviation of Amoro Management Service? It would be helpful to mention the full name at least once.

Author: Good idea, AMS is Amoro Meta Store.

Reviewer 3: I don't see a mention that AMS stands for Amoro Meta Store in the text. Could you please add that to the first time where AMS is mentioned?

Author: already add
